pkg/clustermesh/mcsapi/serviceimport_controller.go

```diff
@@ -0,0 +1,489 @@
+	}
+
+	return ""
+}
+
+func (r *mcsAPIServiceImportReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	nsExists, err := r.doesNamespaceExist(ctx, req)
+	if err != nil {
+		return controllerruntime.Fail(err)
+	}
+	if !nsExists {
+		return controllerruntime.Success()
+	}
+
+	svcImport, svcImportExists, err := r.getSvcImportBase(ctx, req)
+	if err != nil {
+		return controllerruntime.Fail(err)
+	}
+	svcExport, err := r.getSvcExport(ctx, req)
+	if err != nil {
+		return controllerruntime.Fail(err)
+	}
+
+	svcExportByRemoteCluster := r.globalServiceExports.GetServiceExportByCluster(req.NamespacedName)
+	if len(svcExportByRemoteCluster) == 0 && svcExport == nil {
+		if svcImportExists {
+			return controllerruntime.Fail(r.Client.Delete(ctx, svcImport))
+		}
+		return controllerruntime.Success()
+	}
+
+	svcExportByAllCluster := maps.Clone(svcExportByRemoteCluster)
+
+	if svcExport != nil {
+		localSvc, err := r.getLocalService(ctx, req)
+		if err != nil {
+			return controllerruntime.Fail(err)
+		}
+		if localSvc == nil {
+			if meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
+				Type:               mcsapiv1alpha1.ServiceExportValid,
+				Status:             metav1.ConditionFalse,
+				LastTransitionTime: metav1.Now(),
+				Reason:             "NoService",
+				Message:            "Service doesn't exist",
+			}) {
+				if err := r.Client.Status().Update(ctx, svcExport); err != nil {
+					return controllerruntime.Fail(err)
+				}
+			}
+			return controllerruntime.Success()
+		}
+		localSvcSpec := fromServiceToMCSAPIServiceSpec(localSvc, svcExport)
+		if svcExportByAllCluster == nil {
+			svcExportByAllCluster = operator.ServiceExportsByCluster{}
+		}
+		svcExportByAllCluster[r.clusterName] = localSvcSpec
+
+		if localSvc.Spec.Type == corev1.ServiceTypeExternalName {
+			if meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
+				Type:               mcsapiv1alpha1.ServiceExportValid,
+				Status:             metav1.ConditionFalse,
+				LastTransitionTime: metav1.Now(),
+				Reason:             "ServiceType",
+				Message:            "Service type ExternalName is not supported",
+			}) {
+				if err := r.Client.Status().Update(ctx, svcExport); err != nil {
+					return controllerruntime.Fail(err)
+				}
+			}
+			return controllerruntime.Success()
+		}
+	}
+
+	oldestCluster := getOldestRemoteClusterService(svcExportByAllCluster)
+	oldestClusterSvc := svcExportByAllCluster[oldestCluster]
+	mergedPorts := mergePorts(svcExportByAllCluster)
+
+	if svcExport != nil {
+		conflict := false
+		changedCondition := meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
+			Type:               mcsapiv1alpha1.ServiceExportValid,
+			Status:             metav1.ConditionTrue,
+			Reason:             mcsapiv1alpha1.ServiceExportValid,
+			Message:            "Service is Valid for export",
+			LastTransitionTime: metav1.Now(),
+		})
+
+		if oldestCluster != r.clusterName {
```

Comment:
I am not sure if I understand correctly what you mean here, if we are the local source of truth we can't really have a conflict afaict because we are the authority :thinking:. But I might be missing something though

Reply:
Hmm, in my view you'd still have a conflict, although it gets resolved in your favor. But you still have the underlying problem that the services are not aligned across all clusters, which may be worth outlining in the condition, so that it can be easily detected from any cluster (that has the service export). Does that make sense?

Reply:
Ah I see what you mean, yes that make sense and I think is also valuable for admins, so will look at doing this, thanks :eyes:!